theano/compile/function_module.py

```diff
@@ -485,16 +486,178 @@ def __setitem__(self, item, value):
+        storages and output storages with original function
+        """
+        return self.copy()
+
+    def copy(self, share_memory=False, swap=None, delete_updates=False):
+        """
+        Copy this function. Copied function will have separated maker and
+        fgraph with original function. User can choose whether to separate
+        storage by changing the share_memory arguments.
+        Note:
+        We reuse original Out instances but In instances, therefore variables
+        in Ins and Outs are not in the same graph. Please avoid using variables
+        is Ins and Outs to reinitialize another Function or FunctionGraoh.
+        ---------------------
+        Params:
+            share_memory -- { boolean } Default is False. When True, two
+            function share intermediate storages(storages except input and
+            output storages). Otherwise two function will only share partial
+            storages( see method __copy__() ) and same maker. If two functions
+            share memory and allow_gc=False, this will increase executing speed
+            and save memory.
+
+            swap -- { dict } Dictionary<String, theano.SharedVariable> that
+            map old SharedVariable's name to new SharedVariable. Default is
+            None.
+
+            delete_updates -- { boolean } Default is False. If True, Copied
+            function will not have update.
+        ---------------------
+        Returns:
+            func -- Copied theano.Function
+        """
+        maker = self.maker
+        # Copy Ins, so that they have different storage as their value
+        ins, outs = copy.deepcopy([maker.inputs, maker.outputs])
+
+        # Delete update output in fgraph and updates In instancesis needed
+        if delete_updates:
+            # The first len(outs) variabels would be original variables.
+            # The rest variables in fgraph.outputs are the updates.
+            out_vars = maker.fgraph.outputs[:len(outs)]
+        else:
+            out_vars = maker.fgraph.outputs
+        # Init new fgraph using copied variables and get memo
+        # memo: a dict that map old variables to new variabls
+        memo = graph.clone_get_equiv(maker.fgraph.inputs, out_vars)
+        fg_cpy = gof.fg.FunctionGraph([memo[i] for i in maker.fgraph.inputs],
+                                      [memo[o] for o in out_vars],
+                                      clone=False)
+
+        # Swap varaible in Outs.
+        for i in xrange(len(outs)):
+            outs[i].variable = fg_cpy.outputs[i]
+
+        # Swap update and variable in Ins
+        # By doing this, we can pass FunctionMaker._check_unused_inputs()
+        update_i = len(outs)
+        for i, in_var in zip(ins, fg_cpy.inputs):
+            i.variable = in_var
+            if not delete_updates and i.update is not None:
+                i.update = fg_cpy.outputs[update_i]
+                update_i += 1
```

Comment:
if we have an update, but do not have that shared variable as inputs, is that working correctly? Can you make a test with that case?